### PR TITLE
Implement 'Generating' screen

### PR DESCRIPTION
**Pull request by jfquevedol2198**
https://github.com/ioxiocom/ioxio-tags/assets/137529618/b43e52d5-e49a-4e5e-ae44-13083d737911

### CODE REVIEW

**Comment by meandrewandyou** [comment]
input
Input becomes half-disabled on `generating`. Yeah I see in designs it still not grey in this state, but I guess it's the mistake in designs. Can't imagine why should we have checkbox active while other things in the form are disabled

**Comment by meandrewandyou** [comment]
disabled_toggle
Selected half of the toggle doesn't change on `generating`. From the user perspective feels a little bit odd. Maybe it should have it's own disabled state, which a little bit darker then unselected part

**Comment by meandrewandyou** [comment]
not_allowed `not-allowed` cursor appears for buttons, but not for inputs, which are interactive elements too. But maybe there're some conventions of how to use this cursor, Idk

**Comment by meandrewandyou** [comment]
It's possible for now to go to `generating` state with unfilled form. Can't see the separate task for this, but if we already submitting form on this step - maybe it's good moment to add some front end validation - just to check if everything that needs to be filled/selected/checked is actually filled/selected/checked

**Comment by meandrewandyou** [comment]
I thought that was fixed on one of the previous PRs, but on this branch form/description parts still looks like taking around 50/50 of the screen. It looks different in designs
Screenshot 2023-09-28 123020

**Comment by lietu** [comment]
> It's possible for now to go to `generating` state with unfilled form. Can't see the separate task for this, but if we already submitting form on this step

The issue is that this task was done in the wrong order, the API integration should've been first, and it would've included actually submitting the form + basic validation + error handling. It should not do in this.

**Comment by lietu** [comment]
> I thought that was fixed on one of the previous PRs, but on this branch form/description parts still looks like taking around 50/50 of the screen

Agreed with Alina to set it to 50/50 to scale better to larger screens, the discussion for how to tweak it in the future is still open.

File: generator/src/routes/+page.svelte [hunk 1]
```diff
@@ -61,33 +86,54 @@
...
   </div>
   <!-- Right Panel -->
   <div class="qrcode-area-wrapper">
     <div class="qrcode-area">
       <div class="qrcode">
-        <img class="frame" src={SubtractSvg} alt="" aria-hidden="true" />
+        {#if status !== Status.GENERATING}
+          <img class="frame" src={SubtractSvg} alt="" aria-hidden="true" />
+          <img class="logo" src={LogoTagsSvg} alt="" aria-hidden="true" />
+        {:else}
```
Comment (lietu) [inline generator/src/routes/+page.svelte:106]: Why would the logo not be shown during generating? It is in the design.

File: generator/.env.example [hunk 2]
```diff
@@ -0,0 +1,2 @@
+DEVELOPMENT=false
+API_BASE_URL=https://api.tags.ioxio.dev
```
Comment (lietu) [inline generator/.env.example:2]: The default should not point here, it should point to localhost

File: scanner/.env.example [hunk 3]
```diff
@@ -1 +1,2 @@
-DEVELOPMENT=true
+DEVELOPMENT=false
+API_BASE_URL=https://api.tags.ioxio.dev
```
Comment (lietu) [inline scanner/.env.example:2]: The default should not point here, it should point to localhost

File: generator/.env.example [hunk 4]
```diff
@@ -0,0 +1,3 @@
+PUBLIC_DEVELOPMENT=false
+PUBLIC_API_BASE_URL_LOCAL=http://localhost:8081
+API_BASE_URL_PROD=https://api.tags.ioxio.dev
```
Comment (lietu) [inline generator/.env.example:3]: No, this is not how you configure which one you use. You set one variable, it will be `PUBLIC_API_BASE_URL=http://localhost:8081`, and in the deployed environments we will override it.

File: generator/src/routes/+page.svelte [hunk 5]
```diff
@@ -29,30 +98,56 @@
   <div class="form-wrapper">
     <div class="title">Generate a product passport</div>
     <img class="logomarkSvg" src={LogomarkSvg} alt="" aria-hidden="true" />
-    <form>
+    <form method="POST" on:submit|preventDefault={onGenerate}>
```
Comment (lietu) [inline generator/src/routes/+page.svelte:101]: We don't want a `method="POST"`, we're not making POST requests from a form

File: generator/svelte.config.js [hunk 6]
```diff
@@ -27,7 +27,7 @@ const config = {
         "img-src": ["'self'", "data:"],
         "style-src": ["'self'", "'unsafe-inline'"],
         "script-src": ["'self'"],
-        "connect-src": ["'self'", "api.tags.ioxio.dev", "api.tags.ioxio.io"],
+        // "connect-src": ["'self'", "api.tags.ioxio.dev", "api.tags.ioxio.io"],
```
Comment (lietu) [inline generator/svelte.config.js:30]: You don't just remove things like this just because it causes issues, you fix them

File: generator/src/routes/+page.svelte (hunk 1 above)
Comment (jfquevedol2198) [reply generator/src/routes/+page.svelte]: image

**Comment by lietu** [comment]
Split this into 3 different pull requests

File: generator/src/components/Button/index.svelte [hunk 7]
```diff
@@ -1,17 +1,24 @@
 <script lang="ts">
-  export let onClick: () => void
+  export let onClick: (() => void) | null = null
   export let icon: string | null = null
   export let title: string
+  export let disabled: boolean = false
+  export let type: "button" | "submit" | "reset" | null | undefined = "button"
```
Comment (lietu) [inline generator/src/components/Button/index.svelte:6]: Why would we allow `null` or `undefined`?

Comment (lietu) [reply generator/src/components/Button/index.svelte:6]:
```suggestion
  export let type: "button" | "submit" | "reset" = "button"
```

File: generator/src/components/Button/index.svelte [hunk 8]
```diff
@@ -1,17 +1,24 @@
 <script lang="ts">
-  export let onClick: () => void
+  export let onClick: (() => void) | null = null
```
Comment (lietu) [inline generator/src/components/Button/index.svelte:2]: `null` is bad, use `undefined`
```suggestion
  export let onClick: (() => void) | undefined = undefined
```

File: generator/src/components/Button/index.svelte [hunk 9]
```diff
@@ -1,17 +1,24 @@
...
   export let icon: string | null = null
```
Comment (lietu) [inline generator/src/components/Button/index.svelte:3]:
```suggestion
  export let icon: string | undefined = undefined
```

File: generator/src/components/FormCheckbox/index.svelte [hunk 10]
```diff
@@ -1,12 +1,15 @@
 <script lang="ts">
   import Checkbox from "$assets/check.svg?component"
   export let name: string
-  export let label: string = ""
+  export let label: string | undefined
+  export let value: boolean | undefined
```
Comment (lietu) [inline generator/src/components/FormCheckbox/index.svelte:5]: Why would we call this `value` instead of `checked`? Also why would we allow `undefined` for a boolean?

```suggestion
  export let checked: boolean = false
```

File: generator/src/components/FormCheckbox/index.svelte [hunk 11]
```diff
@@ -1,12 +1,15 @@
...
+  export let disabled: boolean | undefined
+  export let required: boolean = false
 </script>
 
 <div class="wrapper">
   <div class="checkbox">
-    <input type="checkbox" id={name} />
+    <input type="checkbox" id={name} {disabled} {name} {required} bind:checked={value} />
```
Comment (lietu) [inline generator/src/components/FormCheckbox/index.svelte:12]:
```suggestion
    <input type="checkbox" id={name} {disabled} {name} {required} bind:checked={checked} />
```

File: generator/src/components/FormCheckbox/index.svelte [hunk 12]
```diff
@@ -1,12 +1,15 @@
 <script lang="ts">
   import Checkbox from "$assets/check.svg?component"
   export let name: string
-  export let label: string = ""
+  export let label: string | undefined
+  export let value: boolean | undefined
+  export let disabled: boolean | undefined
```
Comment (lietu) [inline generator/src/components/FormCheckbox/index.svelte:6]:
```suggestion
  export let disabled: boolean = false
```

File: generator/src/components/FormInputGroup/index.svelte [hunk 13]
```diff
@@ -1,12 +1,14 @@
 <script lang="ts">
   export let name: string
   export let label: string
-  export let placeholder: string = ""
+  export let placeholder: string | undefined
+  export let disabled: boolean | undefined
```
Comment (lietu) [inline generator/src/components/FormInputGroup/index.svelte:5]:
```suggestion
  export let disabled: boolean = false
```

File: generator/src/components/Toggle/index.svelte [hunk 14]
```diff
@@ -1,14 +1,21 @@
 <script lang="ts">
   export let options: string[] = []
   export let value: string
+  export let disabled: boolean | undefined
```
Comment (lietu) [inline generator/src/components/Toggle/index.svelte:4]:
```suggestion
  export let disabled: boolean = false
```

File: generator/src/components/Toggle/index.svelte [hunk 15]
```diff
@@ -1,14 +1,21 @@
...
+
+  function onClickHandler(option: string) {
+    if (disabled) return
```
Comment (lietu) [inline generator/src/components/Toggle/index.svelte:7]:
```suggestion
    if (disabled) {
      return
    }
```

File: generator/src/routes/+page.svelte [hunk 16]
```diff
@@ -6,17 +6,69 @@
...
+  import LogoTagsBlackSvg from "$assets/ioxio-tags-logo-black.svg?url"
   import EffectSvg from "$assets/effect.svg?url"
   import LogomarkSvg from "$assets/ioxio-logomark.svg?url"
   import LogoSvg from "$assets/ioxio-logo.svg?url"
   import QuestionSvg from "$assets/question.svg?component"
   import SubtractSvg from "$assets/subtract.svg?url"
+  import DownloadSvg from "$assets/download.svg?url"
   import type { PageData } from "./$types"
+  import { Status } from "./types"
+  import type { InputData } from "$lib/types"
```
Comment (lietu) [inline generator/src/routes/+page.svelte:18]:
```suggestion
  import type { components } from "$lib/openapi"
  type GenerateSecureV1Request = components["schemas"]["GenerateSecureV1Request"]
```

File: generator/src/lib/types.ts [hunk 17]
```diff
@@ -0,0 +1,6 @@
+export type InputData = {
+  iss: string
+  product: string
+  id: string
+  valid?: boolean
+}
```
Comment (lietu) [inline generator/src/lib/types.ts:6]: This is counterproductive, you have `openapi.ts`. Remove this file.

Comment (lietu) [reply generator/src/lib/types.ts:6]: Additionally the `?` on `valid` is not correct. It is a required field

File: generator/src/routes/+page.svelte [hunk 18]
```diff
@@ -6,17 +6,69 @@
...
+      reader.onload = function () {
+        var imageDataUrl = reader.result
+        if (imageDataUrl) {
+          qrcodeElement.setAttribute("src", imageDataUrl?.toString())
+        }
+      }
+      return
+    }
+    if (result.status === 201) {
+    }
```
Comment (lietu) [inline generator/src/routes/+page.svelte:59]: When does the API return a `201`?

File: generator/src/routes/+page.svelte [hunk 19]
```diff
@@ -6,17 +6,69 @@
...
+    if (result.status === 400) {
+    }
```
Comment (lietu) [inline generator/src/routes/+page.svelte:61]:
```suggestion
    } else {
      // TODO: Better error handling
      console.error("Generating tag failed", result)
    }
```

File: generator/src/routes/+page.svelte [hunk 20]
```diff
@@ -6,17 +6,69 @@
   import Button from "$components/Button/index.svelte"
   import Toggle from "$components/Toggle/index.svelte"
   import LogoTagsSvg from "$assets/ioxio-tags-logo.svg?url"
+  import LogoTagsBlackSvg from "$assets/ioxio-tags-logo-black.svg?url"
```
Comment (lietu) [inline generator/src/routes/+page.svelte:9]:
```suggestion
  import IoxioTagsLogoBlack from "$assets/ioxio-tags-logo-black.svg?url"
```

File: generator/src/routes/+page.svelte [hunk 21]
```diff
@@ -29,30 +81,57 @@
...
+          />
         </div>
         <FormSelectGroup
           name="product"
           label="Product"
           placeholder="Type a product"
           options={data.options}
+          disabled={status === Status.GENERATING}
+          required
         />
```
Comment (lietu) [inline generator/src/routes/+page.svelte:109]: It is not possible for me to enter any product I want. The expected default was "Arbitrary", as in that you can just enter any value you want.

**Comment by lietu** [comment]
Why do you think it is in any way useful for you to put a random background image behind the QR code?

image

I've multiple times stated explicitly that this will be implemented to the backend, and just placing a picture on the background in HTML does not make it appear in the download.

This is also not the correct design.

File: generator/src/routes/+page.svelte [hunk 22]
```diff
@@ -61,33 +140,68 @@
...
-        <img class="frame" src={SubtractSvg} alt="" aria-hidden="true" />
+      <div class="qrcode-frame">
+        {#if status !== Status.GENERATING}
+          <img class="frame" src={SubtractSvg} alt="" aria-hidden="true" />
+          <img class="logo" src={LogoTagsSvg} alt="" aria-hidden="true" />
+          {#if status === Status.FINISHED}
+            <img class="qrcode" alt="" aria-hidden="true" src="" bind:this={qrcodeElement} />
+            <div class="ioxio-tag-frame">
+              <img class="tag" src={LogoTagsBlackSvg} alt="" aria-hidden="true" />
+            </div>
```
Comment (lietu) [inline generator/src/routes/+page.svelte:158]:
```suggestion
            <img class="tag" src={LogoTagsBlackSvg} alt="" aria-hidden="true" />
```

File: generator/src/routes/+page.svelte [hunk 23]
```diff
@@ -61,33 +140,68 @@
...
+            This application serves as a demonstration platform for generating QR codes specifically
+            designed for Digital Product Passports.
+          </p>
+          <p>
+            Utilizing the IOXIO Dataspace technology, this app showcases how QR codes can be
+            generated to serve as digital passports for various Data Products. These QR codes can be
+            read using a compatible reader application, enabling access to relevant product details
+            in real time.
+          </p>
+          <a class="documentation" href="/">See documentation →</a>
```
Comment (lietu) [inline generator/src/routes/+page.svelte:188]: Please put `documentationUrl` in the configuration, there's now multiple links to `/` and finding them later is going to be hard

File: generator/src/routes/+page.svelte [hunk 24]
```diff
@@ -61,33 +140,68 @@
...
+        </div>
+      {:else if status === Status.FINISHED}
+        <div class="result">
+          <p class="label">Issuer domain</p>
+          <p class="value">{inputData.iss}</p>
+          <p class="label">Issuer domain</p>
+          <p class="value">{inputData.product}</p>
+          <p class="label">Issuer domain</p>
+          <p class="value">{inputData.id}</p>
```
Comment (lietu) [inline generator/src/routes/+page.svelte:197]: image

image

File: generator/src/lib/api.ts [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+import type { paths } from "$lib/openapi"
+import { env } from "$env/dynamic/public"
+
+export const API_BASE_URL = env.PUBLIC_API_BASE_URL
+
+const apity = Apity.for<paths>()
+apity.configure({
+  // Base URL to your API
+  baseUrl: API_BASE_URL,
+})
```
[the remark this answers is not in the document]
Comment (lietu) [reply generator/src/lib/api.ts:11]: Actually this is wrong, the configuration AND A DEFAULT VALUE THAT WORKS FOR DEVELOPMENT should be in a config file. I didn't notice earlier but in the `scanner` side this is definitely in the wrong place, this should be in a `config.ts` or `settings.ts` in there too:

image

File: generator/svelte.config.js [hunk 26]
```diff
@@ -27,7 +27,7 @@ const config = {
         "img-src": ["'self'", "data:"],
         "style-src": ["'self'", "'unsafe-inline'"],
         "script-src": ["'self'"],
-        "connect-src": ["'self'", "api.tags.ioxio.dev", "api.tags.ioxio.io"],
+        "connect-src": ["'self'", "api.tags.ioxio.dev", "api.tags.ioxio.io", "localhost:8081"],
```
[the remark this answers is not in the document]
Comment (lietu) [reply generator/svelte.config.js:30]: Ah right I think Denis made it on `scanner`, either way pretty sure this is the format he used

File: generator/src/routes/+page.svelte (hunk 18 above)
Comment (jfquevedol2198) [reply generator/src/routes/+page.svelte]: It should be 422, fault

File: generator/src/lib/api.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (jfquevedol2198) [reply generator/src/lib/api.ts]: Do we need to remove env file, use config file instead of it?

[the remark this answers is not in the document]
Comment (lietu) [reply generator/src/lib/api.ts]: The `.env.example` is just that, an "example", not default. It should stay in the repository, but nobody should need to use it to have the software work for development. When this is deployed on a server I will use env variables to override the configuration, so the `.env.example` and reading from environment is very useful.

Create a `settings.ts` which reads the stuff from `$env/...` into e.g. an 
```ts
export const settings = {
  PUBLIC_API_BASE_URL: env.PUBLIC_API_BASE_URL || "http://localhost:8081",
}
```

File: generator/src/lib/config.ts [hunk 27]
```diff
@@ -1 +1,2 @@
-export const DOCUMENTATION_URL = "/documentation"
+export const DOCUMENTATION_URL: string = "/documentation"
+export const ISS_DOMAIN = ["tags.ioxio.dev"]
```
Comment (lietu) [inline generator/src/lib/config.ts:8]: Why is this an array?

File: generator/src/routes/+page.svelte [hunk 28]
```diff
@@ -61,33 +172,66 @@
...
     <div class="qrcode-area">
-      <div class="qrcode">
-        <img class="frame" src={SubtractSvg} alt="" aria-hidden="true" />
+      <div class="qrcode-frame">
+        {#if status !== Status.GENERATING}
+          <img class="frame" src={SubtractSvg} alt="" aria-hidden="true" />
+          <img class="logo" src={IoxioTagsLogo} alt="" aria-hidden="true" />
+          {#if status === Status.FINISHED}
+            <img class="qrcode" alt="" aria-hidden="true" src="" bind:this={qrcodeElement} />
+            <img class="ioxio-tag-frame" src={IoxioTagsQrcodeFrame} alt="" aria-hidden="true" />
```
Comment (lietu) [inline generator/src/routes/+page.svelte:188]: Please stop adding this, I've *JUST* told you that it will be added by the backend.

File: generator/src/routes/+page.svelte (hunk 24 above)
Comment (lietu) [reply generator/src/routes/+page.svelte:227]: So you resolved the `Issuer domain` being repeated 3x exactly how?

File: scanner/src/routes/+page.server.ts [hunk 29]
```diff
@@ -1,7 +1,7 @@
-import { DEVELOPMENT } from "$env/static/private"
+import { settings } from "$lib/config"
 
 export function load() {
   return {
-    isDevelopment: DEVELOPMENT === "true",
+    isDevelopment: settings.PUBLIC_API_BASE_URL === "true",
```
Comment (lietu) [inline scanner/src/routes/+page.server.ts:5]: Excuse me?

Comment (lietu) [reply scanner/src/routes/+page.server.ts:5]: Please read through your own changes before asking me to review them

File: generator/src/lib/config.ts (hunk 27 above)
Comment (jfquevedol2198) [reply generator/src/lib/config.ts:8]: I misunderstood something about this. Ok its a string now

File: generator/src/lib/config.ts [hunk 30]
```diff
@@ -1,2 +1,8 @@
+import { env } from "$env/dynamic/public"
+
+export const settings = {
+  PUBLIC_API_BASE_URL: env.PUBLIC_API_BASE_URL || "http://localhost:8081",
+}
+
 export const DOCUMENTATION_URL: string = "/documentation"
 export const ISS_DOMAIN = ["tags.ioxio.dev"]
```
[the remark this answers is not in the document]
Comment (jfquevedol2198) [reply generator/src/lib/config.ts:8]: Do yoou want to be like this?
```javascript
export const settings = {
  PUBLIC_API_BASE_URL: env.PUBLIC_API_BASE_URL || "http://localhost:8081",
}

export const config = {
  DOCUMENTATION_URL: "/documentation",
  ISS_DOMAIN: "tags.ioxio.dev"
}
```

File: scanner/src/routes/+page.server.ts (hunk 29 above)
Comment (jfquevedol2198) [reply scanner/src/routes/+page.server.ts:5]: i don't understand what you mean

Comment (jfquevedol2198) [reply scanner/src/routes/+page.server.ts:5]: i don't understand what you mean

File: generator/src/routes/+page.svelte (hunk 24 above)
Comment (jfquevedol2198) [reply generator/src/routes/+page.svelte:227]: What do you mean?

File: scanner/src/routes/+page.server.ts (hunk 29 above)
Comment (jfquevedol2198) [reply scanner/src/routes/+page.server.ts:5]: Its hard to follow your rule because I am just starting to follow the your rule, I hope you give me examples.

Comment (jfquevedol2198) [reply scanner/src/routes/+page.server.ts:5]: Please give me kind explanation

File: generator/src/routes/+page.svelte (hunk 24 above)
Comment (lietu) [reply generator/src/routes/+page.svelte:227]: image

File: generator/src/lib/config.ts (hunk 30 above)
[the remark this answers is not in the document]
Comment (jfquevedol2198) [reply generator/src/lib/config.ts:8]: Ok

File: scanner/src/lib/settings.ts [hunk 31]
```diff
@@ -0,0 +1,6 @@
+import { env } from "$env/dynamic/private"
+
+export const settings = {
+  DEVELOPMENT: true,
```
Comment (lietu) [inline scanner/src/lib/settings.ts:4]: Now this isn't configurable

```suggestion
  DEVELOPMENT: env.DEVELOPMENT === "true",
```

**Review by lietu — APPROVED** [review]
Looks great, thanks!
